BUG: Fix Timestamp constructor changes value on ambiguous DST

AlexKirko opened the pull request:
- [X] closes #24329
- [X] 1 tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

AlexKirko committed:
BUG: fix Timestamp constructor value change on DST Timestamp

AlexKirko commented:
This took some digging, so please bear with me.

**Short version**
When we make a Timestamp using epoch time that corresponds to the second time an ambigous DST time occurs, it gets created and gets assigned a correct `Timestamp.value`. However, when we call the Timestamp constructor again on this Timestamp, this is what happens:

1. `convert_datetime_to_tsobject` in `conversion.pyx` gets called.
2. It eventually calls `npy_datetimestruct_to_datetime` which doesn't care about DST. As it simply adds up date components after converting them to seconds.
3. Then we attempt to shift the `Timestamp.value` using `get_utcoffset`. It sees the state its arguments are in, and, if we use `dateutil` assumes that we are in DST. `pytz` assumes we aren't in DST, which means that only `dateutil` breaks, because we end up shifting `Timezone.value` by DST timedelta (`npy_datetimestruct_to_datetime` doesn't shift, and then `get_utcoffset` shifts).

AlexKirko commented:
Appears like there was a numpy update that changes things for this bug fix. I'll need a day to check this out.

AlexKirko committed:
BUG: expand if check for Linux

AlexKirko commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -382,6 +385,14 @@ cdef _TSObject convert_datetime_to_tsobject(datetime ts, object tz,
+        # Two check in if necessary because class
+        # differs on Windows and Linux
+        if (isinstance(ts.tzinfo, du_tzfile1) or
+                isinstance(ts.tzinfo, du_tzfile2)):
```
This is necessary because ts.tzinfo is `dateutil.zoneinfo.tzfile` on Windows but `dateutil.tz.tz.tzfile` on Linux. I'm not sure why this happens.

AlexKirko commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -382,6 +385,14 @@ cdef _TSObject convert_datetime_to_tsobject(datetime ts, object tz,
+                isinstance(ts.tzinfo, du_tzfile2)):
+            if ts.tzinfo.is_ambiguous(ts):
+                dst_offset = ts.tzinfo.dst(ts)
+                obj.value += int(dst_offset.total_seconds() * 1e9)
```
Because `pydatetime_to_dt64` doesn't take DST into account but `get_utcoffset` does, we need to add the DST timedelta to the result of `pydatetime_to_dt64`. This is necessary only for `dateutil`, because for ambiguous dates `pytz` immediately switches off DST and reduces UTC offset, so no correction is necessary for `pytz` timezones.

jreback replied:
can you add a comment to this effect here

AlexKirko commented on pandas/tests/indexes/datetimes/test_timezones.py:
```diff
@@ -573,13 +573,7 @@ def test_dti_construction_ambiguous_endpoint(self, tz):
-            assert times[-1] == Timestamp("2013-10-27 01:00:00+0100", tz=tz, freq="H")
-        else:
-            assert times[-1] == Timestamp("2013-10-27 01:00:00+0000", tz=tz, freq="H")
+        assert times[-1] == Timestamp("2013-10-27 01:00:00+0000", tz=tz, freq="H")
```
Now no matter whether we use `dateutil` or `pytz` timezones, we get the same date range, so separate testing conditions are no longer necessary.

AlexKirko committed:
CLN: switch from isinstance to treat_tz_asdateutil

AlexKirko commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -5,6 +5,8 @@ cimport numpy as cnp
 from numpy cimport int64_t, int32_t, intp_t, ndarray
 cnp.import_array()
 
+from dateutil.tz import tzfile as _dateutil_tzfile
```
This is used in `timezones.pyx` when we need to check for `dateutil` timezone and make sure it works both on Windows and Linux. Seems to work.

mroeschke commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -382,6 +384,13 @@ cdef _TSObject convert_datetime_to_tsobject(datetime ts, object tz,
+        # GH 24329 Take DST offset into account
+        # use dateutil.tz.tzfile to check type
+        if ts.tzinfo is not None:
+            if isinstance(ts.tzinfo, _dateutil_tzfile):
```
Think you can use`treat_tz_as_dateutil`  https://github.com/pandas-dev/pandas/blob/d78c061669b7c1a0a50a393a8e58d660d84a79f9/pandas/_libs/tslibs/timezones.pyx#L42

mroeschke replied:
And can also remove the `if ts.tzinfo is not None` check if you use `treat_tz_as_dateutil` directly.

mroeschke commented:
Does this also fix the case for nonexistent times near DST?

https://github.com/pandas-dev/pandas/issues/24329#issuecomment-506546068

AlexKirko commented:
@mroeschke Unfortunately, no. Shifts it back by an hour. I would have to look into why this happens.

AlexKirko commented:
@mroeschke Okay, looked into it. You'll be happy to know that, according to `dateutil`, summer time starts 128 nanoseconds before 2 A.M. Looks to be a bug in `dateutil` that isn't connected to the issue that this PR is trying to handle. The bug manifests starting with 128 seconds before 2 A.M. both on master and my branch.
**Details**
This is fine:
```
>>> pd.__version__
'0.26.0.dev0+1773.g664d928fd'
>>> epoch =  1552211999999999871
>>> t = pd.Timestamp(epoch, tz='dateutil/US/Pacific')
>>> t
Timestamp('2019-03-10 01:59:59.999999871-0800', tz='dateutil/US/Pacific')
>>> t.value
1552211999999999871
>>> pd.Timestamp(t)
Timestamp('2019-03-10 01:59:59.999999871-0800', tz='dateutil/US/Pacific')
>>> pd.Timestamp(t).value
1552211999999999871
```
This is also fine:
```
>>> epoch =  1552212000000000000
>>> t = pd.Timestamp(epoch, tz='dateutil/US/Pacific')
>>> t
Timestamp('2019-03-10 03:00:00-0700', tz='dateutil/US/Pacific')
>>>
>>> t.value
1552212000000000000
>>> pd.Timestamp(t)
Timestamp('2019-03-10 03:00:00-0700', tz='dateutil/US/Pacific')
>>> pd.Timestamp(t).value
1552212000000000000
```
Meanwhile, this breaks representation and gets us nonexistent times:
```
>>> epoch =  1552211999999999872
>>> t = pd.Timestamp(epoch, tz='dateutil/US/Pacific')
>>> t
Timestamp('2019-03-10 01:59:59.999999872-0700', tz='dateutil/US/Pacific')
>>> t.value
1552211999999999872
>>> pd.Timestamp(t)
Timestamp('2019-03-10 01:59:59.999999872-0800', tz='dateutil/US/Pacific')
>>> pd.Timestamp(t).value
1552208399999999872
```
And right on the cusp, the value breaks too:
```
>>> epoch =  1552211999999999999
>>> t = pd.Timestamp(epoch, tz='dateutil/US/Pacific')
>>> t
Timestamp('2019-03-10 01:59:59.999999999-0700', tz='dateutil/US/Pacific')
>>> t.value
1552211999999999999
>>> pd.Timestamp(t)
Timestamp('2019-03-10 01:59:59.999999999-0800', tz='dateutil/US/Pacific')
>>> pd.Timestamp(t).value
1552208399999999999
```
So the bug is there, but it turns out to be unrelated and matters only for 128 nanoseconds before the switch to summer time.
I can make a separate issue, but frankly, I don't know what we are supposed to do with this at `pandas` level. Seems wrong to add crutches to our code because of a `dateutil` bug that almost no one will ever come across.

AlexKirko commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -382,6 +382,11 @@ cdef _TSObject convert_datetime_to_tsobject(datetime ts, object tz,
     else:
         obj.value = pydatetime_to_dt64(ts, &obj.dts)
+        # GH 24329 Take DST offset into account
+        if treat_tz_as_dateutil(ts.tzinfo):
```
@mroeschke Switched to `treat_tz_as_dateutil`. Much cleaner now, thank you.

mroeschke reviewed (approved):
Nice investigative work! LGTM.

Yes, please make a separate issue for the nonexistent cases. Your examples are really good.

mroeschke commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -968,6 +968,7 @@ Datetimelike
 - Bug in :func:`date_range` with custom business hours as ``freq`` and given number of ``periods`` (:issue:`30593`)
 - Bug in :class:`PeriodIndex` comparisons with incorrectly casting integers to :class:`Period` objects, inconsistent with the :class:`Period` comparison behavior (:issue:`30722`)
 - Bug in :meth:`DatetimeIndex.insert` raising a ``ValueError`` instead of a ``TypeError`` when trying to insert a timezone-aware :class:`Timestamp` into a timezone-naive :class:`DatetimeIndex`, or vice-versa (:issue:`30806`)
+- Bug in :class:`Timestamp` where constructing :class:`Timestamp` from ambiguous epoch time and calling constructor again changed :meth:`Timestamp.value` property (:issue:`24329`)
```
Sorry one more thing, can you move to whatsnew version 1.1.0?

jreback reviewed (approved):
minor comments, otherwise lgtm (ex moving the whatsnew to 1.1). @mroeschke pls merge when satisfied.

AlexKirko committed:
CLN: add comment to solution and move to v1.1.0

AlexKirko committed:
DOC: reword bugfix comment

AlexKirko commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -362,6 +362,14 @@ cdef _TSObject convert_datetime_to_tsobject(datetime ts, object tz,
+        # GH 24329 When datetime is ambiguous,
+        # pydatetime_to_dt64 doesn't take DST into account
+        # but with dateutil timezone, get_utcoffset does
+        # so we need to correct for it
```
Added a comment explaining the reasoning for this solution. Tried to make it as brief as possible.

AlexKirko commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -59,6 +59,7 @@ Categorical
 
 Datetimelike
 ^^^^^^^^^^^^
+- Bug in :class:`Timestamp` where constructing :class:`Timestamp` from ambiguous epoch time and calling constructor again changed :meth:`Timestamp.value` property (:issue:`24329`)
```
Moved to whatsnew for v1.1.0

AlexKirko commented:
@mroeschke Made the required changes. Also performed the final cleanup: rebased to updated master and squashed tiny commits.
Please take a look if it's ready for merging.

mroeschke commented:
Thanks @AlexKirko!

If you'd like to continue investigating https://github.com/pandas-dev/pandas/issues/31043, it would be much appreciated.
